libraries/AP_RangeFinder/AP_RangeFinder_Lua.cpp

```diff
@@ -22,20 +22,49 @@
 
 // constructor
 AP_RangeFinder_Lua::AP_RangeFinder_Lua(RangeFinder::RangeFinder_State &_state, AP_RangeFinder_Params &_params) :
-    AP_RangeFinder_Backend(_state, _params)
+    AP_RangeFinder_Backend(_state, _params), _distance_m(0.0f), _signal_quality_pct(-1)
 {
     set_status(RangeFinder::Status::NoData);
 }
 
 
 // Set the distance based on a Lua Script
 bool AP_RangeFinder_Lua::handle_script_msg(float dist_m)
+{
+    return handle_script_msg_ex(dist_m, -1.0f);
+}
+
+
+// Set the distance and signal_quality based on a Lua Script.
+// If the signal_quality_pct argument is an int8_t from lua, then
+// things don't work. Not sure why but using a float works better.
+bool AP_RangeFinder_Lua::handle_script_msg_ex(float dist_m, float signal_quality_pct)
 {
     state.last_reading_ms = AP_HAL::millis();
     _distance_m = dist_m;
+
+    // Trying to set a bogus signal_quality is like having no signal_quality data.
+    _signal_quality_pct = (signal_quality_pct >= 0.0f && signal_quality_pct <= 100.0f) ?
+            static_cast<int8_t>(signal_quality_pct) : -1;
     return true;
 }
 
+bool AP_RangeFinder_Lua::get_signal_quality_pct(int8_t &quality_pct) const
+{
+    // Status and distance are updated via a periodic update routine. signal_quality
+    // is returned via this overridden virtual method. It seems that there is
+    // a good chance that the distance measurement and the signal_quaqlity value can
+    // get out of sync. For example if a hardware driver collects both distance and
+    // signal_quality and then a client queries the distance before the update routine
+    // runs, then the client will get and old distance with a new signal_quality.
+    // signal_quality should really be in the state structure along with distance.
```

Comment:
I agree. Perhaps folks would be open to adding a field to RangeFinder::RangeFinder_State. Worth discussing.